lib/widgets/content.dart

```diff
@@ -382,6 +387,188 @@ class MessageImage extends StatelessWidget {
   }
 }
 
+class MessageVideo extends StatelessWidget {
+  const MessageVideo({super.key, required this.node});
+
+  final VideoNode node;
+
+  @override
+  Widget build(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+
+    // For YouTube and Vimeo links, display a widget with a thumbnail.
+    // When the thumbnail is tapped, open the video link in an external browser or
+    // a supported external app.
+    if (node.previewImageUrl != null) {
+      return MessageEmbedVideoPreview(
+        src: node.srcUrl, previewImage: node.previewImageUrl!);
+    }
+
+    final resolvedSrc = store.tryResolveUrl(node.srcUrl);
+    return resolvedSrc != null
+      ? MessageInlineVideoPreview(src: resolvedSrc)
+      : Container();
+  }
+}
+
+class MessageEmbedVideoPreview extends StatelessWidget {
+  const MessageEmbedVideoPreview({
+    super.key,
+    required this.previewImage,
+    required this.src,
```

Comment:
Looking at these two widget classes for the two kinds of videos, and at `MessageVideo` which dispatches to them, I think it'd probably be best to have two different ContentNode subclasses for them.

In particular there's a very important difference in semantics between these two, if I'm correctly understanding how these work:
* For "inline" videos, `node.srcUrl` is a URL on the Zulip server, and so it's fine to load it when the user just scrolls passively past that message (and doing so is helpful for the reasons you mention in the PR description).
* For "embed" videos, `node.srcUrl` is a URL on some *other* external service. For privacy reasons it's therefore important that we not go and talk to that service unless the user actually chooses to interact with the video. OTOH `node.previewImageUrl` is a URL back on the user's Zulip server (provided by Camo), and safe to eagerly fetch from.

I believe this PR already handles that distinction correctly. But in this version it's pretty subtle in the code, and so it'd be easy for some future change to break that privacy property without realizing there was something important there. We should therefore arrange the code so that the distinction is clear. As part of that, we should avoid having these two kinds of URLs with very different privacy properties both live as values of the same field on the same class.

Conversely, `MessageVideo` is doing very little, really just dispatching to these two related classes that have pretty different logic from each other. So even apart from the privacy issue, it feels like having two different `ContentNode` subclasses would better match up with how the logic actually works.

If there's anyplace where it would be helpful to do so, those two `ContentNode` subclasses could always have a common base class with a name like `VideoNode`.